File: app/src/main/java/com/baijum/ukufretboard/ui/ProgressionPracticeView.kt
package com.baijum.ukufretboard.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowLeft
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.FilterChip
import androidx.compose.material3.FilterChipDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Slider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateMapOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.baijum.ukufretboard.audio.MetronomeEngine
import com.baijum.ukufretboard.data.ChordFormulas
import com.baijum.ukufretboard.data.Notes
import com.baijum.ukufretboard.data.Progression
import com.baijum.ukufretboard.data.VoicingGenerator
import com.baijum.ukufretboard.domain.ChordVoicing
import com.baijum.ukufretboard.domain.Note
import com.baijum.ukufretboard.viewmodel.FretboardViewModel
import com.baijum.ukufretboard.viewmodel.UkuleleString

/**
 * Full-screen practice view for a chord progression.
 *
 * Shows the progression chords in a timeline, a fretboard diagram that
 * updates to the current chord, a voicing selector with left/right arrows,
 * and playback controls (BPM, beats per chord, loop, play/stop).
 *
 * Accessed from the Progressions tab via a "Practice" action on each card,
 * following the same content-replacement pattern as VoiceLeadingView.
 *
 * @param progression The chord progression to practice.
 * @param keyRoot The root pitch class of the selected key.
 * @param tuning Current ukulele tuning for voicing generation and fretboard.
 * @param lastFret The highest fret shown on the fretboard (from user settings).
 * @param leftHanded Whether to mirror diagrams for left-handed players.
 * @param onPlayVoicing Callback to play a single chord voicing (null if sound disabled).
 * @param onBack Callback to exit the practice view.
 * @param modifier Optional modifier.
 */
@Composable
fun ProgressionPracticeView(
    progression: Progression,
    keyRoot: Int,
    tuning: List<UkuleleString>,
    lastFret: Int = 12,
    leftHanded: Boolean = false,
    onPlayVoicing: ((ChordVoicing) -> Unit)? = null,
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val scope = rememberCoroutineScope()
    val engine = remember { MetronomeEngine() }

    // Playback state
    var isPlaying by remember { mutableStateOf(false) }
    var bpm by remember { mutableFloatStateOf(100f) }
    var beatsPerChord by remember { mutableIntStateOf(4) }
    var loop by remember { mutableStateOf(true) }
    var currentChordIndex by remember { mutableIntStateOf(0) }

    // All voicings per chord degree (list of lists)
    val allVoicings: List<List<ChordVoicing>> = remember(progression, keyRoot, tuning) {
        progression.degrees.map { degree ->
            val chordRoot = (keyRoot + degree.interval) % Notes.PITCH_CLASS_COUNT
            val formula = ChordFormulas.ALL.firstOrNull { it.symbol == degree.quality }
            if (formula != null) {
                VoicingGenerator.generate(chordRoot, formula, tuning)
            } else {
                emptyList()
            }
        }
    }

    // Selected voicing index per chord position (defaults to 0)
    val selectedVoicingIndex = remember { mutableStateMapOf<Int, Int>() }

    // Helper to get the selected voicing for a chord index
    fun voicingFor(chordIdx: Int): ChordVoicing? {
        val voicings = allVoicings.getOrNull(chordIdx) ?: return null
        if (voicings.isEmpty()) return null
        val idx = (selectedVoicingIndex[chordIdx] ?: 0).coerceIn(0, voicings.lastIndex)
        return voicings[idx]
    }

    // Current voicing for the displayed chord
    val currentVoicing = voicingFor(currentChordIndex)
    val currentVoicingCount = allVoicings.getOrNull(currentChordIndex)?.size ?: 0
    val currentVoicingIdx = (selectedVoicingIndex[currentChordIndex] ?: 0)
        .coerceIn(0, (currentVoicingCount - 1).coerceAtLeast(0))

    // Build selections map for FretboardView from the current voicing
    // Muted strings (MUTED = -1) map to null (unplayed)
    val selections: Map<Int, Int?> = if (currentVoicing != null) {
        currentVoicing.frets.mapIndexed { stringIdx, fret ->
            stringIdx to (if (fret == ChordVoicing.MUTED) null else fret)
        }.toMap()
    } else {
        tuning.indices.associateWith { null }
    }

    // getNoteAt for the fretboard
    val getNoteAt: (Int, Int) -> Note = { stringIndex, fret ->
        val pc = (tuning[stringIndex].openPitchClass + fret) % Notes.PITCH_CLASS_COUNT
        Note(pitchClass = pc, name = Notes.pitchClassToName(pc))
    }

    // Clean up metronome on dispose
    DisposableEffect(Unit) {
        onDispose {
            engine.stop()
        }
    }

    Column(
        modifier = modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(top = 8.dp),
    ) {
        // ── Header: Back + progression name ──
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            IconButton(
                onClick = {
                    engine.stop()
                    isPlaying = false
                    onBack()
                },
            ) {
                Icon(
                    imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                    contentDescription = "Back",
                )
            }
            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = "Practice",
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                Text(
                    text = progression.name,
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.Bold,
                )
            }
            // Key label
            Text(
                text = "Key: ${Notes.enharmonicForKey(keyRoot, keyRoot)}",
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.primary,
                fontWeight = FontWeight.Bold,
                modifier = Modifier.padding(end = 8.dp),
            )
        }

        Spacer(modifier = Modifier.height(12.dp))

        // ── Chord timeline ──
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .horizontalScroll(rememberScrollState())
                .padding(horizontal = 16.dp),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            progression.degrees.forEachIndexed { index, degree ->
                val chordRoot = (keyRoot + degree.interval) % Notes.PITCH_CLASS_COUNT
                val chordName = Notes.enharmonicForKey(chordRoot, keyRoot) + degree.quality
                val isCurrent = index == currentChordIndex

                Box(
                    modifier = Modifier
                        .background(
                            color = if (isCurrent) {
                                MaterialTheme.colorScheme.primary
                            } else {
                                MaterialTheme.colorScheme.surfaceVariant
                            },
                            shape = RoundedCornerShape(10.dp),
                        )
                        .clickable(enabled = !isPlaying) {
                            currentChordIndex = index
                        }
                        .padding(horizontal = 14.dp, vertical = 10.dp),
                    contentAlignment = Alignment.Center,
                ) {
                    Column(horizontalAlignment = Alignment.CenterHorizontally) {
                        Text(
                            text = chordName,
                            style = MaterialTheme.typography.bodyMedium,
                            fontWeight = if (isCurrent) FontWeight.Bold else FontWeight.Normal,
                            color = if (isCurrent) {
                                MaterialTheme.colorScheme.onPrimary
                            } else {
                                MaterialTheme.colorScheme.onSurfaceVariant
                            },
                        )
                        Text(
                            text = degree.numeral,
                            style = MaterialTheme.typography.labelSmall,
                            color = if (isCurrent) {
                                MaterialTheme.colorScheme.onPrimary.copy(alpha = 0.8f)
                            } else {
                                MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.7f)
                            },
                        )
                    }
                }
            }
        }

        Spacer(modifier = Modifier.height(12.dp))

        // ── Fretboard ──
        FretboardView(
            tuning = tuning,
            selections = selections,
            showNoteNames = true,
            onFretTap = { _, _ -> }, // read-only
            getNoteAt = getNoteAt,
            leftHanded = leftHanded,
            lastFret = lastFret,
            cellWidth = 48.dp,
            cellHeight = 42.dp,
            scrollable = true,
        )

        Spacer(modifier = Modifier.height(8.dp))

        // ── Voicing selector ──
        if (currentVoicingCount > 0) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 16.dp),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.Center,
            ) {
                IconButton(
                    onClick = {
                        val cur = selectedVoicingIndex[currentChordIndex] ?: 0
                        val newIdx = if (cur > 0) cur - 1 else currentVoicingCount - 1
                        selectedVoicingIndex[currentChordIndex] = newIdx
                    },
                    enabled = currentVoicingCount > 1,
                    modifier = Modifier.size(36.dp),
                ) {
                    Icon(
                        imageVector = Icons.AutoMirrored.Filled.KeyboardArrowLeft,
                        contentDescription = "Previous voicing",
                    )
                }

                Text(
                    text = "Voicing ${currentVoicingIdx + 1} of $currentVoicingCount",
                    style = MaterialTheme.typography.bodyMedium,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.width(140.dp),
                )

                IconButton(
                    onClick = {
                        val cur = selectedVoicingIndex[currentChordIndex] ?: 0
                        val newIdx = if (cur < currentVoicingCount - 1) cur + 1 else 0
                        selectedVoicingIndex[currentChordIndex] = newIdx
                    },
                    enabled = currentVoicingCount > 1,
                    modifier = Modifier.size(36.dp),
                ) {
                    Icon(
                        imageVector = Icons.AutoMirrored.Filled.KeyboardArrowRight,
                        contentDescription = "Next voicing",
                    )
                }
            }
        } else {
            Text(
                text = "No voicing available",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center,
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(8.dp),
            )
        }

        Spacer(modifier = Modifier.height(12.dp))

        // ── Controls card ──
        Card(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 16.dp),
            shape = RoundedCornerShape(12.dp),
            colors = CardDefaults.cardColors(
                containerColor = MaterialTheme.colorScheme.secondaryContainer,
            ),
        ) {
            Column(modifier = Modifier.padding(12.dp)) {
                // BPM slider + Tap Tempo
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Text(
                        text = "BPM: ${bpm.toInt()}",
                        style = MaterialTheme.typography.labelSmall,
                        modifier = Modifier.width(64.dp),
                    )
                    Slider(
                        value = bpm,
                        onValueChange = { bpm = it },
                        valueRange = 40f..220f,
                        steps = 0,
                        modifier = Modifier.weight(1f),
                    )
                    TapTempoButton(
                        onBpmDetected = { detected -> bpm = detected.toFloat() },
                        modifier = Modifier.padding(start = 4.dp),
                    )
                }

                Spacer(modifier = Modifier.height(4.dp))

                // Beats per chord
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(6.dp),
                ) {
                    Text(
                        text = "Beats:",
                        style = MaterialTheme.typography.labelSmall,
                    )
                    listOf(1, 2, 4, 8).forEach { beats ->
                        FilterChip(
                            selected = beatsPerChord == beats,
                            onClick = { beatsPerChord = beats },
                            label = { Text("$beats") },
                            colors = FilterChipDefaults.filterChipColors(
                                selectedContainerColor = MaterialTheme.colorScheme.primary,
                                selectedLabelColor = MaterialTheme.colorScheme.onPrimary,
                            ),
                            modifier = Modifier.height(30.dp),
                        )
                    }
                }

                Spacer(modifier = Modifier.height(8.dp))

                // Loop + Play / Stop
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(12.dp),
                ) {
                    // Loop toggle
                    FilterChip(
                        selected = loop,
                        onClick = { loop = !loop },
                        label = { Text("Loop") },
                        leadingIcon = {
                            Icon(
                                Icons.Filled.Refresh,
                                contentDescription = "Toggle loop",
                                modifier = Modifier.size(16.dp),
                            )
                        },
                        colors = FilterChipDefaults.filterChipColors(
                            selectedContainerColor = MaterialTheme.colorScheme.primary,
                            selectedLabelColor = MaterialTheme.colorScheme.onPrimary,
                            selectedLeadingIconColor = MaterialTheme.colorScheme.onPrimary,
                        ),
                        modifier = Modifier.height(30.dp),
                    )

                    Spacer(modifier = Modifier.weight(1f))

                    // Play / Stop
                    IconButton(
                        onClick = {
                            if (isPlaying) {
                                engine.stop()
                                isPlaying = false
                            } else {
                                isPlaying = true
                                var lastChordPlayed = -1
                                engine.start(
                                    scope = scope,
                                    bpm = bpm.toInt(),
                                    beatsPerChord = beatsPerChord,
                                    chordCount = progression.degrees.size,
                                    loop = loop,
                                    onBeat = { chordIdx, _ ->
                                        currentChordIndex = chordIdx
                                        if (chordIdx != lastChordPlayed) {
                                            lastChordPlayed = chordIdx
                                            val voicing = voicingFor(chordIdx)
                                            if (voicing != null && onPlayVoicing != null) {
                                                onPlayVoicing(voicing)
                                            }
                                        }
                                    },
                                    onComplete = {
                                        isPlaying = false
                                    },
                                )
                            }
                        },
                        modifier = Modifier
                            .size(40.dp)
                            .background(
                                color = if (isPlaying) {
                                    MaterialTheme.colorScheme.error
                                } else {
                                    MaterialTheme.colorScheme.primary
                                },
                                shape = CircleShape,
                            ),
                    ) {
                        Icon(
                            imageVector = if (isPlaying) Icons.Filled.Close else Icons.Filled.PlayArrow,
                            contentDescription = if (isPlaying) "Stop" else "Play",
                            tint = if (isPlaying) {
                                MaterialTheme.colorScheme.onError
                            } else {
                                MaterialTheme.colorScheme.onPrimary
                            },
                        )
                    }
                }
            }
        }

        Spacer(modifier = Modifier.height(16.dp))
    }
}

File: iosApp/UkuleleCompanion/Views/ProgressionPracticeView.swift
import SwiftUI
import Combine
import shared

struct ProgressionPracticeView: View {
    let progression: Progression
    let keyRoot: Int32

    @State private var currentChordIndex = 0
    @State private var selectedVoicingIndex: [Int: Int] = [:]
    @State private var bpm: Float = 100
    @State private var beatsPerChord = 4
    @State private var loop = true
    @State private var isPlaying = false
    @State private var timer: AnyCancellable?

    private let tonePlayer = TonePlayer()
    private let leftHanded: Bool
    private let tuning: [shared.UkuleleString]
    private let cachedVoicings: [[ChordVoicing]]

    private static let sampleNames = [
        "uke_c", "uke_csharp", "uke_d", "uke_dsharp",
        "uke_e", "uke_f", "uke_fsharp", "uke_g",
        "uke_gsharp", "uke_a", "uke_asharp", "uke_b",
    ]

    init(progression: Progression, keyRoot: Int32) {
        self.progression = progression
        self.keyRoot = keyRoot
        let defaults = UserDefaults(suiteName: "app_settings") ?? .standard
        self.leftHanded = defaults.bool(forKey: "left_handed")
        let t = UkuleleTuning.highG
        let tuning = t.asUkuleleStrings
        self.tuning = tuning

        let degrees = progression.degrees.asArray(of: ChordDegree.self)
        let formulas = ChordFormulas.shared.ALL.asArray(of: ChordFormula.self)
        self.cachedVoicings = degrees.map { degree in
            let chordRoot = (keyRoot + degree.interval) % 12
            guard let formula = formulas.first(where: { $0.symbol == degree.quality }) else {
                return []
            }
            let voicings = VoicingGenerator.shared.generate(
                rootPitchClass: chordRoot,
                formula: formula,
                tuning: tuning,
                allowMutedStrings: false
            )
            return voicings.asArray(of: ChordVoicing.self)
        }
    }

    private func voicingFor(_ chordIdx: Int) -> ChordVoicing? {
        guard chordIdx < cachedVoicings.count, !cachedVoicings[chordIdx].isEmpty else { return nil }
        let idx = min(selectedVoicingIndex[chordIdx] ?? 0, cachedVoicings[chordIdx].count - 1)
        return cachedVoicings[chordIdx][max(0, idx)]
    }

    var body: some View {
        let voicings = cachedVoicings
        let currentVoicing = voicingFor(currentChordIndex)
        let currentVoicingCount = currentChordIndex < voicings.count ? voicings[currentChordIndex].count : 0
        let currentVoicingIdx = min(selectedVoicingIndex[currentChordIndex] ?? 0, max(0, currentVoicingCount - 1))
        let degrees = progression.degrees.asArray(of: ChordDegree.self)

        ScrollView {
            VStack(spacing: 12) {
                // Header
                VStack(spacing: 4) {
                    Text("Practice")
                        .font(.caption)
                        .foregroundStyle(.secondary)
                    Text(progression.name)
                        .font(.headline)
                    Text("Key: \(Notes.shared.enharmonicForKey(pitchClass: keyRoot, keyRoot: nil, isMinor: false))")
                        .font(.subheadline)
                        .foregroundStyle(Color.accentColor)
                }
                .padding(.top, 8)

                // Chord timeline
                ScrollView(.horizontal, showsIndicators: false) {
                    HStack(spacing: 8) {
                        ForEach(0..<degrees.count, id: \.self) { index in
                            let degree = degrees[index]
                            let chordRoot = (keyRoot + degree.interval) % 12
                            let chordName = Notes.shared.enharmonicForKey(
                                pitchClass: chordRoot, keyRoot: nil, isMinor: false
                            ) + degree.quality
                            let isCurrent = index == currentChordIndex

                            Button {
                                if !isPlaying {
                                    currentChordIndex = index
                                }
                            } label: {
                                VStack(spacing: 2) {
                                    Text(chordName)
                                        .font(.subheadline.bold())
                                    Text(degree.numeral)
                                        .font(.caption2)
                                        .opacity(0.8)
                                }
                                .padding(.horizontal, 14)
                                .padding(.vertical, 10)
                                .background(isCurrent ? Color.accentColor : Color(.systemGray5))
                                .foregroundStyle(isCurrent ? .white : .primary)
                                .clipShape(RoundedRectangle(cornerRadius: 10))
                            }
                            .buttonStyle(.plain)
                            .accessibilityAddTraits(isCurrent ? .isSelected : [])
                        }
                    }
                    .padding(.horizontal)
                }

                // Chord diagram
                if let voicing = currentVoicing {
                    let degree = degrees[currentChordIndex]
                    let chordRoot = (keyRoot + degree.interval) % 12
                    let chordName = Notes.shared.enharmonicForKey(
                        pitchClass: chordRoot, keyRoot: nil, isMinor: false
                    ) + degree.quality
                    ChordDiagramView(voicing: voicing, chordName: chordName)
                        .frame(maxWidth: .infinity)
                } else {
                    Text("No voicing available")
                        .font(.subheadline)
                        .foregroundStyle(.secondary)
                        .padding(.vertical, 32)
                }

                // Voicing selector
                if currentVoicingCount > 0 {
                    HStack {
                        Button {
                            let cur = selectedVoicingIndex[currentChordIndex] ?? 0
                            selectedVoicingIndex[currentChordIndex] = cur > 0 ? cur - 1 : currentVoicingCount - 1
                        } label: {
                            Image(systemName: "chevron.left")
                        }
                        .disabled(currentVoicingCount <= 1)
                        .accessibilityLabel("Previous voicing")

                        Text("Voicing \(currentVoicingIdx + 1) of \(currentVoicingCount)")
                            .font(.subheadline)
                            .frame(width: 140)

                        Button {
                            let cur = selectedVoicingIndex[currentChordIndex] ?? 0
                            selectedVoicingIndex[currentChordIndex] = cur < currentVoicingCount - 1 ? cur + 1 : 0
                        } label: {
                            Image(systemName: "chevron.right")
                        }
                        .disabled(currentVoicingCount <= 1)
                        .accessibilityLabel("Next voicing")
                    }
                }

                // Controls card
                controlsCard(degrees: degrees)
            }
            .padding(.bottom, 16)
        }
        .navigationTitle("Practice")
        .navigationBarTitleDisplayMode(.inline)
        .onDisappear { stopPlayback() }
        .onChange(of: currentChordIndex) { newIndex in
            guard isPlaying, newIndex < degrees.count else { return }
            let degree = degrees[newIndex]
            let chordRoot = (keyRoot + degree.interval) % 12
            let name = Notes.shared.enharmonicForKey(
                pitchClass: chordRoot, keyRoot: nil, isMinor: false
            ) + degree.quality
            AccessibilityAnnouncer.shared.announce(name, minInterval: 0.5)
        }
    }

    // MARK: - Controls Card

    private func controlsCard(degrees: [ChordDegree]) -> some View {
        VStack(spacing: 8) {
            // BPM slider
            HStack {
                Text("BPM: \(Int(bpm))")
                    .font(.caption)
                    .frame(width: 64, alignment: .leading)
                Slider(value: $bpm, in: 40...220)
                    .accessibilityValue("\(Int(bpm)) beats per minute")
                Button("Tap") { tapTempo() }
                    .font(.caption.bold())
                    .padding(.horizontal, 8)
                    .padding(.vertical, 4)
                    .background(Color(.systemGray5))
                    .clipShape(Capsule())
            }

            // Beats per chord
            HStack(spacing: 6) {
                Text("Beats:")
                    .font(.caption)
                ForEach([1, 2, 4, 8], id: \.self) { beats in
                    Button {
                        beatsPerChord = beats
                    } label: {
                        Text("\(beats)")
                            .font(.caption.bold())
                            .padding(.horizontal, 12)
                            .padding(.vertical, 4)
                            .background(beatsPerChord == beats ? Color.accentColor : Color(.systemGray5))
                            .foregroundStyle(beatsPerChord == beats ? .white : .primary)
                            .clipShape(Capsule())
                    }
                    .buttonStyle(.plain)
                }
            }

            // Loop + Play/Stop
            HStack {
                Button {
                    loop.toggle()
                } label: {
                    Label("Loop", systemImage: "repeat")
                        .font(.caption.bold())
                        .padding(.horizontal, 12)
                        .padding(.vertical, 6)
                        .background(loop ? Color.accentColor : Color(.systemGray5))
                        .foregroundStyle(loop ? .white : .primary)
                        .clipShape(Capsule())
                }
                .buttonStyle(.plain)

                Spacer()

                Button {
                    if isPlaying {
                        stopPlayback()
                    } else {
                        startPlayback(chordCount: degrees.count)
                    }
                } label: {
                    Image(systemName: isPlaying ? "stop.fill" : "play.fill")
                        .font(.title3)
                        .foregroundStyle(.white)
                        .frame(width: 40, height: 40)
                        .background(isPlaying ? Color.red : Color.accentColor)
                        .clipShape(Circle())
                }
                .accessibilityLabel(isPlaying ? "Stop" : "Play")
            }
        }
        .padding()
        .background(Color(.secondarySystemBackground))
        .clipShape(RoundedRectangle(cornerRadius: 12))
        .padding(.horizontal)
    }

    // MARK: - Playback

    private func startPlayback(chordCount: Int) {
        guard chordCount > 0 else { return }
        isPlaying = true
        currentChordIndex = 0
        var beatCount = 0
        var chordIdx = 0
        var lastChordPlayed = -1

        let interval = 60.0 / Double(bpm)
        timer = Timer.publish(every: interval, on: .main, in: .common)
            .autoconnect()
            .sink { _ in
                if chordIdx != lastChordPlayed {
                    lastChordPlayed = chordIdx
                    if let voicing = voicingFor(chordIdx) {
                        playVoicing(voicing)
                    }
                }
                currentChordIndex = chordIdx
                beatCount += 1

                if beatCount >= beatsPerChord {
                    beatCount = 0
                    chordIdx += 1
                    if chordIdx >= chordCount {
                        if loop {
                            chordIdx = 0
                        } else {
                            stopPlayback()
                        }
                    }
                }
            }
    }

    private func stopPlayback() {
        timer?.cancel()
        timer = nil
        tonePlayer.stop()
        isPlaying = false
    }

    private func playVoicing(_ voicing: ChordVoicing) {
        let frets = voicing.fretInts
        for (i, fret) in frets.enumerated() {
            guard fret >= 0 else { continue }
            let pc = (tuning[i].openPitchClass + Int32(fret)) % 12
            tonePlayer.play(Self.sampleNames[Int(pc)])
            return
        }
    }

    // MARK: - Tap Tempo

    @State private var tapTimestamps: [Date] = []

    private func tapTempo() {
        let now = Date()
        tapTimestamps.append(now)
        // Keep last 4 taps
        if tapTimestamps.count > 4 {
            tapTimestamps.removeFirst()
        }
        // Reset if gap > 2 seconds
        if tapTimestamps.count >= 2 {
            let last = tapTimestamps[tapTimestamps.count - 1]
            let prev = tapTimestamps[tapTimestamps.count - 2]
            if last.timeIntervalSince(prev) > 2.0 {
                tapTimestamps = [now]
                return
            }
        }
        guard tapTimestamps.count >= 2 else { return }
        let intervals = zip(tapTimestamps.dropFirst(), tapTimestamps).map { $0.timeIntervalSince($1) }
        let avg = intervals.reduce(0, +) / Double(intervals.count)
        let detected = 60.0 / avg
        bpm = Float(min(220, max(40, detected)))
    }
}
